Fmt: optimize format rule of let_binder, if_stmt, [x..y]

trdthg opened the pull request:
- Optimizing format rule of `let_binder`, `if_stmt`, `[x.... .y]` ... and their combinations 
- if_stmt formatting rule change: if any of [i, t, e] wrap, then both [t, e] force to wrap
- add more nowrap rule
	- atom
	- string_literal
	- field
	- app
	- binary
	- index
	- if_stmt if all of [i, t, e] can be nowrap
	- trailing comment
	- delim

I also tested it on sail-riscv/model/*.sail

You can see the formatting results compared to sail2 branch at https://github.com/trdthg/sail-riscv/pull/2 

There's still a lot of work to be done, but that will be next time

> and share a blog I read a few days ago:: https://laurent.le-brun.eu/blog/the-story-of-reformatting-100k-files-at-google-in-2011

arichardson commented:
I had another look at the diff and a lot of it is definitely an improvement, but there is a bit of weirdness with more complex expressions inside the `if` condition.

trdthg commented:
Update

1. Allow `Infix_sequence` nowrap, this will fix things like
	```ocaml
	if
        i > end_element | i >= real_num_elem
      then {
	```

2. 

```
a + b + c + d
[][][][][]
f(f(f(f(f(f(f(f(), f()), f()))))))
if f(a, f(a, b)) then { f(a, f(f(a, b), f(a, b))) } else { .................................... }
```

all these can be collapsed into one line now

I put some restrictions on purpose before.

For example, a maximum of two levels of nesting `f(f())` to avoid situations that would make a line too long like `f(f(f(f(f(f()))))))))`. But it's hard to quantify.

And I found that you can configure `max_width` in _sail_config.json_, and ocaml PPrint will automatically handle line breaks. So, I've removed these restrictions now.

trdthg commented:
Update:

1. force exps in toplevel function block wrap

	```diff
	-function clause execute SINVAL_VMA(rs1, rs2) = { execute(SFENCE_VMA(rs1, rs2)) }
	+function clause execute SINVAL_VMA(rs1, rs2) = {
	+  execute(SFENCE_VMA(rs1, rs2))
	+}
	```

trdthg commented:
Update:

- Allow Struct, Tuple, Unary nowrap
- Allow binary_rhs(when block) no prefix hardline
	
	i.e.

	```
	let a = match {
		...
	}

arichardson commented:
Can't comment on the OCaml code, but the resulting sail formatting is IMO a big improvement.

Mostly unrelated note: Some of the one-line if/else clauses are a bit complicated and I'd personally add line breaks or maybe set the column limit lower. Is there a way to force multiline if/else? Maybe there should be a config option to say don't merge if/else into a single line unless it's part of a let/other expression?

So basically this is good in one line: `let trace_size = if rvfi_int_data_present then { trace_size + 320 } else { trace_size }`
But maybe 
```
function hex_bits_signed_forwards bv = {
  if signed(bv) < 0 then { (length(bv), concat_str("-", hex_str(unsigned(not_vec(bv) + 1)))) } else {
    (length(bv), hex_str(unsigned(bv)))
  }
}
```

should be 

```
function hex_bits_signed_forwards bv = {
  if signed(bv) < 0 then {
    (length(bv), concat_str("-", hex_str(unsigned(not_vec(bv) + 1))))
  } else {
    (length(bv), hex_str(unsigned(bv)))
  }
}
```

trdthg commented:
Thank you for your review. I agree, in fact the [comment](https://github.com/rems-project/sail/pull/636#issuecomment-2254049711) above mentions a restriction, and it was used to solve this issue (but didn't work well at last)

> don't merge if/else into a single line unless it's part of a let/other expression

This seems to be good and should solve the problem here, I'll give it a try

```ocaml
function hex_bits_signed_forwards bv = {
  if signed(bv) < 0 then { (length(bv), concat_str("-", hex_str(unsigned(not_vec(bv) + 1)))) } else {
    (length(bv), hex_str(unsigned(bv)))
  }
}
```

The reason it's formatted this way now is that this line is wider than 120, so `PPrint` insert linebreak after `else {` 

If there's a way to make it do the same after `then {`, that would be nice too

Anyway, I'll try both.

trdthg committed:
Fmt: optimize fmt for let_binder, exps, if_stmt ...
- new module
  - WrapChecker: check if `chunk[s]` must wrap
- new function
  - doc_chunks_nowrap_or_surround: check nowrap, if not, then surround
  - doc_chunks_rhs: check wrap, if true then prefix hardline
- format rule updates
  - if_stmt will be prefix a hardline and nested if wrap
  - force foreach wrap

trdthg committed:
Fix `Seq.iteri` is not avaliale in old version ocaml

trdthg commented:
May need to refactor, so turning it into a Draft for now
